irmin-layered: New fancy stats for freeze

Ngoguey42 opened the pull request:
```
freeze 0 blocked 40.672s (35%), and yielded 1min15s (65%). Total 1min56s.
  Event counts: copy_contents:208847, copy_nodes:1465555, copy_commits:13315, copy_branches:0, adds:1312338, skips:2369230, yields:13316, copy_newies_loops:1
  Longest block 3.454s (during "copy newies (loop)"). Longest yield 1min15s (during "copy to lower"). 114.231 yield per sec.
  Timeline:
    wait for freeze lock took 1.27us (0% of total) and blocked 0.873us (0% of total).
                    misc took 471us (0% of total) and blocked 471us (0% of total).
           copy to lower took 1min20s (69% of total) and blocked 4.661s (11% of total).
      copy to next upper took 0.299us (0% of total) and blocked 0.291us (0% of total).
           copy branches took 3.791us (0% of total) and blocked 3.792us (0% of total).
             flush lower took 131us (0% of total) and blocked 131us (0% of total).
      copy newies (loop) took 36.114s (31% of total) and blocked 35.999s (89% of total).
     wait for batch lock took 1.765us (0% of total) and blocked 0.671us (0% of total).
      copy newies (last) took 65.661us (0% of total) and blocked 65.646us (0% of total).
                    misc took 11.78ms (0% of total) and blocked 11.821ms (0% of total).
                finalize took 358us (0% of total) and blocked 317us (0% of total).

```

icristescu commented on bench/irmin-pack/layers.ml:
```diff
@@ -60,35 +56,15 @@ let checkout_and_commit config repo c nb =
-          t.waiting_freeze
-          Fmt.(Dump.list float)
-          t.completed_freeze !total);
-    total := 0)
```
we want to count the number of objects added in upper before each freeze, to have an estimate of the how many objects freeze has to copy. So we need to print and reset `total` at each freeze.

Ngoguey42 replied:
see latest version

icristescu commented on src/irmin-layers/stats.ml:
```diff
@@ -15,113 +15,187 @@
+  mutable outside_totlen : float;
+  mutable inside_totlen : float;
+  mutable outside_maxlen : string * float;
+  mutable inside_maxlen : string * float;
```
it would be nice if we kept here all blocking portion that last more than 1s.

Ngoguey42 replied:
Very good idea

icristescu commented on src/irmin-layers/stats.ml:
```diff
@@ -15,113 +15,187 @@
+  let c = ref 0 in
+  fun () ->
+    incr c;
+    !c - 1
```
can't you start from `-1`?

icristescu commented on src/irmin-layers/stats.ml:
```diff
@@ -15,113 +15,187 @@
 
-(** Ensure lists are not growing indefinitely by dropping elements. *)
-let limit_length_list = 10
+let freeze_start t0 current_section =
```
you can replace `current_section` here with "wait for freeze lock"

Ngoguey42 replied:
I renamed the parameter to `initial_section`. I like having all the section names hard-coded in a single file, i.e. ext_layered.ml

icristescu commented on src/irmin-pack/layered/ext_layered.ml:
```diff
@@ -755,34 +759,24 @@ struct
+    Irmin_layers.Stats.freeze_section "copy to lower";
     (if t.X.Repo.with_lower then
      let min = if squash then max else min in
-     with_stats "copied in lower" (Copy.CopyToLower.copy ~cancel t ~min max)
```
this print allowed us to estimate the number of objects copied to lower vs the ones copied to upper. We don't necessarily need to have this info all the time; but it would be nice to have a way to only count the objects copied in lower for instance, if we need that at some point. (We can also do this in a separate PR, as its not immediately useful).

Ngoguey42 replied:
see latest version

icristescu commented on src/irmin-pack/layered/ext_layered.ml:
```diff
@@ -827,65 +814,54 @@ struct
-              Field.span "add-lock" waiting_add;
-            ])
+      Irmin_layers.Stats.freeze_stop ();
+      (* Fmt.pr "\n%a%!" Irmin_layers.Stats.pp_latest (); *)
```
looks like you forgot to remove this (and few other prints in stats.ml)

CraigFe commented on bench/irmin-pack/layers.ml:
```diff
@@ -60,35 +56,15 @@ let checkout_and_commit config repo c nb =
-          t.completed_freeze !total);
-    total := 0)
+let print_stats () =
+  Logs.app (fun l -> l "%a%!" Irmin_layers.Stats.pp_latest ())
```
```suggestion
  Logs.app (fun l -> l "%t" Irmin_layers.Stats.pp_latest)
```
Flushing isn't necessary inside `Logs`, which takes care of adding necessary flushes / newlines, and you can use `%t` to avoid passing unnecessary unit arguments to pretty-printers (w/ corresponding change to `pp_latest` below).

Ngoguey42 replied:
I forgot to remove that flush, and I didn't know about `%t`, thanks!

CraigFe commented on src/irmin-layers/stats.ml:
```diff
@@ -15,113 +15,187 @@
-  stats_i.commits <- 0;
-  stats_i.branches <- 0
+let freeze_profiles = ref []
+let latest = ref (fresh_freeze_profile ~-1 (Mtime_clock.now ()) "")
```
```suggestion
let latest = ref (fresh_freeze_profile (-1) (Mtime_clock.now ()) "")
```

CraigFe commented on src/irmin-layers/stats.ml:
```diff
@@ -15,113 +15,187 @@
-    skips = stats_t.skips;
-  }
+  freeze_profiles := [];
+  latest := fresh_freeze_profile ~-1 (Mtime_clock.now ()) ""
```
```suggestion
  latest := fresh_freeze_profile (-1) (Mtime_clock.now ()) ""
```

CraigFe commented on src/irmin-layers/stats.ml:
```diff
@@ -15,113 +15,187 @@
-  | `Waiting -> stats_t.waiting_freeze <- span :: stats_t.waiting_freeze
+  freeze_profiles := shorter (!latest :: !freeze_profiles)
+
+let pp_latest ppf () =
```
```suggestion
let pp_latest ppf =
```

(...and use `%t` at callsites rather than `%a`.)

CraigFe commented on src/irmin-test/layered_store.ml:
```diff
@@ -358,17 +358,8 @@ module Make_Layered (S : LAYERED_STORE) = struct
-          (concat_list s.copied_branches)
-          s.nb_freeze)
+  let log_stats () =
+    Logs.debug (fun l -> l "%a%!" Irmin_layers.Stats.pp_latest ())
```
```suggestion
    Logs.debug (fun l -> l "%t" Irmin_layers.Stats.pp_latest)
```

gs0510 commented on src/irmin-layers/stats.mli:
```diff
@@ -14,65 +14,62 @@
+val get_copied_contents_count : unit -> int
+val get_copied_nodes_count : unit -> int
+val get_freeze_count : unit -> int
+val pp_latest : Format.formatter -> unit -> unit
```
There is perhaps an extra `unit` here, the declaration doesn't match the implementation. :)

Ngoguey42 replied:
Yup, I've just fixed it locally :)

Ngoguey42 commented:
New display:
- I've fixed the printing behavior with ongoing freezes. @gs0510 
- Recorded the long blocks/yields
- Addressed all your comments except 2

A finished freeze when `minimum_seconds_to_be_considered_long` is set to 1s.
```
freeze 0 (finished) blocked 38.128s (34%), and yielded 1min15s (66%). Total 1min53s. 117.364 yield per sec.
  Event counts: copy_contents:208847, copy_nodes:1465540, copy_commits:13315, copy_branches:0, adds:1312337, skips:2613006, yields:13316, copy_newies_loops:7
  Longests blocks: [1 long block in "copy newies (loop)" of 2.942s; 1 long block in "copy to lower" of 2.747s].
  Longest yield: 471ms (during "copy to lower").
  Timeline:
    wait for freeze lock took 1.304us (0% of total) and blocked 0.74us (0% of total).
                    misc took 379us (0% of total) and blocked 380us (0% of total).
           copy to lower took 12.612s (11% of total) and blocked 3.914s (10% of total).
      copy to next upper took 0.215us (0% of total) and blocked 0.208us (0% of total).
           copy branches took 4.344us (0% of total) and blocked 4.344us (0% of total).
             flush lower took 77.862us (0% of total) and blocked 77.883us (0% of total).
      copy newies (loop) took 1min40s (89% of total) and blocked 34.204s (90% of total).
     wait for batch lock took 1.633us (0% of total) and blocked 0.818us (0% of total).
      copy newies (last) took 89.125us (0% of total) and blocked 89.141us (0% of total).
                    misc took 9.89ms (0% of total) and blocked 9.891ms (0% of total).
                finalize took 309us (0% of total) and blocked 307us (0% of total).
```

An ongoing freeze when `minimum_seconds_to_be_considered_long` is set to 10ms. See the way long blocks are displayed @icristescu
```
freeze 0 (ongoing) blocked 18.82s (32%), and yielded 39.17s (68%). Total 57.996s. 126.491 yields per sec.
  Event counts: copy_contents:138369, copy_nodes:899337, copy_commits:7336, copy_branches:0, adds:803688, skips:1242940, yields:7336, copy_newies_loops:4
  Longests blocks: [53 long blocks in "copy newies (loop)" of ~90.151ms (max:2.758s); 4 long blocks in "copy to lower" of ~559ms].
  Longests yields: [304 long yields in "copy newies (loop)" of ~22.65ms; 39 long yields in "copy to lower" of ~38.28ms (max:395ms)].
  Timeline:
    wait for freeze lock took 1.483us (0% of total) and blocked 1.012us (0% of total).
                    misc took 286us (0% of total) and blocked 286us (0% of total).
           copy to lower took 9.757s (17% of total) and blocked 3.114s (17% of total).
      copy to next upper took 0.193us (0% of total) and blocked 0.189us (0% of total).
           copy branches took 3.487us (0% of total) and blocked 3.5us (0% of total).
             flush lower took 83.441us (0% of total) and blocked 83.522us (0% of total).
      copy newies (loop) is ongoing.
```

Ngoguey42 commented:
I've pushed the one version to rule them all
- counters are now per-section
- the number of adds preceding the freeze are shown in the first line
- we can see the stats of an ongoing section

```
freeze 0 (ongoing) blocked 6.753s (40%), and yielded 10.111s (60%). Total 16.871s. 265077 adds before freeze. Copy newies loops: 1.
  Longests blocks: [1 long block in "copy newies (loop)" of 2.505s; 1 long block in "copy to lower" of 2.072s].
  Longest yield: 551ms (during "copy to lower").
  Timeline timings:
    wait for freeze lock took 1.411us (0% of total) and blocked 0.734us (0% of total).
                    misc took 260us (0% of total) and blocked 260us (0% of total).
           copy to lower took 10.1s (60% of total) and blocked 3.096s (46% of total).
      copy to next upper took 0.521us (0% of total) and blocked 0.513us (0% of total).
           copy branches took 3.91us (0% of total) and blocked 3.91us (0% of total).
             flush lower took 124us (0% of total) and blocked 124us (0% of total).
      copy newies (loop) took 6.77s (40% of total) and blocked 3.664s (54% of total) (ongoing).
  Timeline counters:
           copy to lower copy_contents:39355, copy_nodes:220133, copy_commits:1665, copy_branches:0, adds:122798, skips:81013, skip_tests:342166, yields:1665.
      copy newies (loop) copy_contents:43895, copy_nodes:234313, copy_commits:672, copy_branches:0, adds:52209, skips:106060, skip_tests:385932, yields:672 (ongoing).
```

icristescu reviewed (approved):
The new stats look great, thanks!

Ngoguey42 commented:
I'm going to merge this one then
